serverlessworkflow/modules/ROOT/pages/core/understanding-jq-expressions.adoc

```diff
@@ -1 +1,159 @@
-//= Understanding JQ expressions
\ No newline at end of file
+= Understanding JQ expressions
+
+Every workflow instance is associated with a data model. This model regardless of whether your flow file is written in YAML or JSON, consists of a JSON object. The initial content of that object depends on how the flow is started. If the flow is created through a link:{cloud_events_url}[Cloud Event], then the content is taken from `data` property. If started through an HTTP POST invocation, then the request body is expected to have a `workflowdata` property, from where the content of the model is retrieved. 
+
+Expressions are the mechanism defined by the link:{spec_doc_url}#workflow-expressions[Serverless Workflow Specification] for the states to interact with the data model. Kogito supports two expression languages: link:https://github.com/json-path/JsonPath[jsonpath] and link:https://stedolan.github.io/jq/manual/[jq]. The default one is `jq`, but you can change it to `jsonpath` by using the `expressionLang` property. This guide will focus on `jq` and discuss its usage in following uses cases: switch state conditions, action function arguments, and data filtering.
+
+
+== Switch Conditions
+
+Conditions inside a switch state allow the workflow designer to choose which path the flow should follow depending on the model content. A condition is an expression that returns a boolean when evaluated against the model. If the condition associated with a state transition returns true, that is the place for the flow to go. 
+
+For example, in link:{kogito_sw_examples_url}/serverless-workflow-greeting-quarkus[greetings] repository, we are selecting which message should be displayed to the user depending on his language of choice: English or Spanish. In computational terms, if the value of the property language is English, the constant literal to be injected on the property `message` will be __Hello from__… , else if the value of the same property is Spanish, then the injected `message` will be __Saludos desde…__.
+
+image::core/switch_condition.png[]
+
+The switch state contains this condition, which in turn contains two `jq` expression that returns a boolean. 
+
+[source,json]
+----
+"dataConditions": [
+        {
+          "condition": "${ .language == \"English\" }",
+          "transition": "GreetInEnglish"
+        },
+        {
+          "condition": "${ .language == \"Spanish\" }",
+          "transition": "GreetInSpanish"
+        }
+      ]
+----
+
+As the specification requires, all expressions are embedded within `${… }`. But Kogito is smart enough to infer that the string within the condition is an expression, so you can skip it and just write a shorter version of the previous code that behaves the same. 
```

Comment:
Guys, @ricardozanini @krisv , @fjtirado  I'm not sure this simplification is a good idea.

Let me explain the benefits and drawbacks I see:

**Benefits:**

- Yo don't have to write the intial "${"  and the  final "}"

**Drawbacks:**

- By one side we claim that we implement the SW standard, all right, however the first thing we learn people here is to not use the standard.

- I think this adds an extra complexity for people to understand the examples, etc. For example, If you missed to read exactly this paragraph of this particular guide, and you read the SW spec, and you jump to the SW kogito-examples, you'll find expressions that you won't understand.

- Consistency, this opens the door to have SWs where in some expressions you use the  ${}, and in other expression you don't.  Even when it can be argued that it's users choice, etc, I don't believe it'll look good, and increases the chance of the problems described in the point above, difficult the migration to another vendor, etc.

- Considering that we implement the SW standard, ideally a SW  you run in kogito (if you don't custom services or other custom stuff) should run in other vendors implementation. This expressions will never do. So we are basically are adding an additional unnecessary incompatibility.

Reply:
We are telling people that if they find the standard too verbose they can skip the ${ (which by the way, was pretty controversial when defininf the standard)
Obviously, if they are worried about being 100% standard compliant, they have to use ${}
Honestly, I do nod see whats the problem here.

Reply:
In my opinion the guides and the examples should use the standard, for example this expression  "operation": "{max: .numbers | max_by(.x), min: .numbers | min_by(.y)}" is not in the standard.

The reasons are more or less explained above, etc, but think that as part of the learning procedure, developers normally read the spec, jump to the examples, go back to the spec, etc.  And as I commented above, unless you have read this particular paragraph of this particular guide, people will find expressions that they won't understand. Doesn't help the learning procedure, etc.

The controversial matters around the "$" I think are a separate discussion, but right now the standard is what it is.

Reply:
If a user is able to understand "operation": "${{max: .numbers | max_by(.x), min: .numbers | min_by(.y)}}" Im pretty sure they will also able to understand "operation": "{max: .numbers | max_by(.x), min: .numbers | min_by(.y)}"

Reply:
I think we can bring this discussion to another place, where we can have a more dynamic conversation about this subject.